File: moderate/robo_and_robitta.rb
File.open(ARGV[0]).each_line do |line|
  r = 0
  s = line.scan(/([0-9]+)x([0-9]+) \| ([0-9]+) ([0-9]+)/)[0].map(&:to_i)
  while s[1] > s[3]
    r += s[0]
    s = s[1] - 1, s[0], s[1] - s[3], s[2]
  end
  puts r + s[2]
end

File: moderate/robo_and_robitta.pl
use strict;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    $line =~ /(\d+)x(\d+) \| (\d+) (\d+)/;
    my ( $m, $n, $s, $t ) = ( $1, $2, $3, $4 );
    my $r = 0;
    while ( $n > $t ) {
        $r += $m;
        ( $m, $n, $s, $t ) = ( $n - 1, $m, $n - $t, $s );
    }
    printf( "%d\n", $r + $s );
}
close(INFILE);
